Spigot-Server-Patches/0425-Use-result-of-setItemStack-in-PlayerBucketEmptyEvent.patch

```diff
@@ -0,0 +1,53 @@
+         }
+     }
+ 
++    public void setItemInHand(EnumHand enumhand, ItemStack itemstack) { this.a(enumhand, itemstack); } // Paper - OBFHELPER
+     public void a(EnumHand enumhand, ItemStack itemstack) {
+         if (enumhand == EnumHand.MAIN_HAND) {
+             this.setSlot(EnumItemSlot.MAINHAND, itemstack);
+diff --git a/src/main/java/net/minecraft/server/ItemBucket.java b/src/main/java/net/minecraft/server/ItemBucket.java
+index 0ff92aea5..498cd4d24 100644
+--- a/src/main/java/net/minecraft/server/ItemBucket.java
++++ b/src/main/java/net/minecraft/server/ItemBucket.java
+@@ -134,12 +134,13 @@ public class ItemBucket extends Item {
+             } else {
+                 // CraftBukkit start
+                 if (entityhuman != null) {
+-                    PlayerBucketEmptyEvent event = CraftEventFactory.callPlayerBucketEmptyEvent(world, entityhuman, blockposition, clicked, enumdirection, itemstack, enumhand);
++                    PlayerBucketEmptyEvent event = CraftEventFactory.callPlayerBucketEmptyEvent(world, entityhuman, blockposition, clicked, enumdirection, entityhuman.isCreative() ? entityhuman.getItemInHand(enumhand) : itemstack, enumhand); // Paper - 2651: Prevents creative players from having their buckets modified
+                     if (event.isCancelled()) {
+                         ((EntityPlayer) entityhuman).playerConnection.sendPacket(new PacketPlayOutBlockChange(world, blockposition)); // SPIGOT-4238: needed when looking through entity
+                         ((EntityPlayer) entityhuman).getBukkitEntity().updateInventory(); // SPIGOT-4541
+                         return false;
+                     }
++                    entityhuman.setItemInHand(enumhand, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItemStack())); // Paper - 2651: set item in hand if event isn't cancelled
+                 }
+                 // CraftBukkit end
+                 if (world.worldProvider.isNether() && this.fluidType.a(TagsFluid.WATER)) {
+diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+index c0fd2948e..aed1cca44 100644
+--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
++++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+@@ -377,7 +377,7 @@ public class CraftEventFactory {
+     }
+ 
+     public static PlayerBucketEmptyEvent callPlayerBucketEmptyEvent(World world, EntityHuman who, BlockPosition changed, BlockPosition clicked, EnumDirection clickedFace, ItemStack itemstack, EnumHand enumHand) {
+-        return (PlayerBucketEmptyEvent) getPlayerBucketEvent(false, world, who, changed, clicked, clickedFace, itemstack, Items.BUCKET, enumHand);
++        return (PlayerBucketEmptyEvent) getPlayerBucketEvent(false, world, who, changed, clicked, clickedFace, itemstack, itemstack.getItem(), enumHand); // Paper - don't forcefully set to bucket
```

Comment:
fixes the creative case but breaks survival - only use the item in hand Item if the player is creative, else Bucket will do

Reply:
should be fixed in latest commit